### ISSUE TITLE
[FEATURE_REQUEST] Group-dependent content (dynamic)

### ISSUE THREAD

**Issue by vogelfreiheit** [Lissy93/dashy#342]
### Is your feature request related to a problem? If so, please describe.

With Keycloak and similar systems you can return user group information as well to use in your application. While it is not always trivial to implement it, it can come handy to dynamically adjust the content presented to the user. In Dashy this could be used to decide if certain items are rendered. For example, NOC/network ops admins, vs plain users, HR, .... It could make the project quite useful for intranet dashboard applications that use LDAP/Keycloak/AD and co.

### Describe the solution you'd like

A method, ex. using specific fields in the content definition, to dynamically adjust the items presented to users based upon their affiliation with specific groups, post-authentication.

I could help with some advice on how to safely retrieve or process the membership information with LDAP and co. I'm not knowledgeable with these fancypants javascript-everything frameworks of today (I can still run my web apps on devices with 32MB worth of memory!), so that would take some learning to do.

A caveat emptor: this depends on proper web server configuration. Keycloak might be the easiest and is already supported. 

### Priority

Medium (Would be very useful)

### Is this something you would be keen to implement

Maybe

**Comment by Lissy93** [comment]
Thanks for the feature request, will look into it :)
A similar option is currently supported, called Granular access, where only certain sections are visible to certain users. But that's only supported with Basic Auth atm, but should be quite straight-forward for me to extend that functionality to KeyCloak, providing a static user ID or some kind of reference can be returned :)

**Comment by vogelfreiheit** [comment]
Hi,

Take a look at this:
https://www.keycloak.org/docs/latest/securing_apps/

What you essentially need is SAML 2.0 attributes. This should be supported out of the box. I will look into it, but feel free to setup a testing env for FreeIPA or Keycloak alone. If you combine both you also learn how to do sysadmin duties of RHEL's IDM solution, so that might come handy for you in the future. It isn't as daunting as it looks.

This has some nice examples that seem applicable:
https://ordina-jworks.github.io/security/2019/08/22/Securing-Web-Applications-With-Keycloak.html

TL;DR:
- You need to be able to authenticate and retrieve some user attributes in a token.
- This probably needs some internal parsing in your application (you can retrieve the token data from the Keycloak server, the token is pinned in the request headers).
- Cache/session store the attributes you use. You may also want to sanitize the groups to alphanumeric strings just for peace of mind.

-

**Comment by AndersBennedsgaard** [comment]
What is the status of this request? I'm working on a Kubernetes cluster with multiple tenants using KeyCloak for authentication, and this would be a nice addition to our toolstack 😃

**Comment by Lissy93** [comment]
Am working on some other tasks right now, and have a lot on at the moment, but will try and implement this sometime in the next few weeks. In the meantime, if you'd like to pick  this up, that would be awesome. Feel free to reach out if you need any support :)

**Comment by Lissy93** [comment]
This is now implemented by @leocov-dev in #397 🙌 
Thank you Leo :)
